### PR TITLE
SWPBL-188190 -- Draw block palette

### PR DESCRIPTION
**Pull request by KevinKimSonos**
## Overview

The block palette is the portion of the GUI that contains all types of Control/DSP blocks that can be used when creating a DSP system. The block palette should be anchored on the left side of the application window beneath the menu bar.

## Notes

There are several things that we need to confirm.
1. In order for the palette to be under the menu bar, I set the palette's initial y-position to be `ImGui::GetFrameHeight() * 2`. This looks fine on my screen, but need to check if it renders what I intended in other screens.
2. Notice that  _Search_ is weirdly placed next to the search bar. This can be fixed by changing a line in _imgui.cpp_. However, I am having trouble understanding how I can edit _imgui.cpp_.
3. Notice that when clicked on the canvas editor screen, the palette gets hidden in the back. This needs to be fixed using the _docking_ feature of ImGui.

***
See JIRA [here](https://jira.sonos.com/browse/SWPBL-188190).

### CODE REVIEW

File: example/Dynamite/palette.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        "smooth_multi_gain",
+        "soft_clipper",
+        "sonar_spatial_processing",
+        "squancher",
+        "thermal_estimator",
+        "tone_control",
+        "weighted_sum",
+        "xfade_ramp_iir"
+    };
+    static std::vector<std::string> dsp_blocks(dsp_list, dsp_list + 29);
```
Comment (mattmooresonos) [inline example/Dynamite/palette.cpp:124]: you can just initialize the static vector of strings

```
static std::vector<std::string> v = { "xyzzy", "plugh", "abracadabra" };
```
and maybe you could have a private member function std::vector<std::string>&listDSPBlocks() that could scope the static vector. This way when we get the python tool integrated we can have it return the same list?

File: example/Dynamite/palette.cpp [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    // Control blocks
+    std::string control_list[] =
```
Comment (mattmooresonos) [inline example/Dynamite/palette.cpp:127]: you can just initialize the static vector of strings

static std::vector<std::string> v = { "xyzzy", "plugh", "abracadabra" };
and maybe you could have a private member function std::vectorstd::string&listControlBlocks() that could scope the static vector. This way when we get the python tool integrated we can have it return the same list?

File: example/Dynamite/palette.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                            ImGui::Indent();
+                            ImGui::TreePop();
+                        }
+                    }
+                    ImGui::TreePop();
+                }				
+            }
+        }
+        ImGui::ListBoxFooter();
+    }
```
Comment (mattmooresonos) [inline example/Dynamite/palette.cpp:183]: can we move this into another private function this one is getting pretty big

**Review by mattmooresonos — APPROVED** [review]
This is looking pretty good! I like the direction we are taking. Some quick edits I think we should make before moving on to the next.

**Comment by EvanShenkman-Sonos** [comment]
Seeing some weird behavior when expanding/collapsing the sections of the block list. Expanding/collapsing one (i.e. IO Blocks) will collapse/expand another (i.e. Control Blocks)

**Comment by KevinKimSonos** [comment]
I will keep a note of Evan's comment in our design document and fix the problem before the end of this sprint.

**Pull request merged**